### PR TITLE
Feature/pandas api sem

### PR DESCRIPTION
**Pull request by tortolavivo23**
# Feature sem

-  Solves: [/issues/17](https://github.com/hablapps/pykx/issues/17)

## What does this change introduce?

- Provides an implemementation for the `sem` method

**Parameters:**

| Name         | Type | Description                                                                      | Default |
| :----------: | :--: | :------------------------------------------------------------------------------- | :-----: |
| axis         | int  | The axis to calculate the sum across 0 is columns, 1 is rows.                    | 0       |
| skipna       | bool | not yet implemented                                           | True    |
| numeric_only | bool | Only use columns of the table that are of a numeric data type.                   | False   |
| ddof    | int  | Delta Degrees of Freedom. The divisor used in calculations is N - ddof, where N represents the number of elements. | 1 |

**Returns:**

| Type               | Description                                                          |
| :----------------: | :------------------------------------------------------------------- |
| Dictionary         | The sem across each row / column with the key corresponding to the row number or column name. |

It follow the [pandas definition](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sem.html). It returns a dictionary instead of a pandas series or a pandas dataframe.

## General

- [x] Has an example been added to demo the new feature?
- [ ] Have existing examples been updated or tested?
- [ ] Have you added any new Environment Variables/Configuration Options? If yes please tick the boxes below as applicable
    - [ ] Addition to reimporter logic within `src/pykx/pykx.q` and `src/pykx/reimporter.py`
    - [ ] Have updated the `src/pykx/util.py` logic which is used for environment variable
- [ ] If there have been any dependency updates have they been reflected in all files?
    - [ ] pyproject.toml
    - [ ] docs/getting-started/installing.md
    - [ ] conda-recipe/meta.yaml
    - [ ] README.md
- [ ] If any examples have been updated has it's associated `.zip` been updated

## Code

- [x] Has all temporary code used during development been removed?
- [x] Has all commented out (unused) code been removed?
- [x] Where reasonable have you ensured there is no duplication of existing code?
- [x] If applicable for your use-case have you ensured that the code is performant?

## Testing

- [x] Have unit tests been created or existing ones updated to test this new functionality?

## Documentation

- [x] Has documentation been added for all public code?
- [ ] Has a release note been included for the new feature?
- [ ] Has any documentation which would benefit from this feature been updated to use the most up to date functionality?
- [ ] If a new class has been added has a documentation stub `.md` file associated with it been created?
- [ ] If any documentation page has been created has it been added to `mkdocs.yml`
- [x] Have you checked your changes with a spell checker? (US English)

### CODE REVIEW

File: docs/user-guide/advanced/Pandas_API.ipynb [hunk 1]
```diff
@@ -646,6 +646,108 @@
...
+   "source": [
+    "tab.sem(axis=1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ae7afe5a",
+   "metadata": {},
+   "source": [
+    "Calculate std accross columns with ddof=0:"
```
Comment (nipsn) [inline docs/user-guide/advanced/Pandas_API.ipynb:738]: We are not calculating the std here, right?

File: docs/user-guide/advanced/Pandas_API.ipynb [hunk 2]
```diff
@@ -646,6 +646,108 @@
...
+    "| axis         | int  | The axis to calculate the sum across 0 is columns, 1 is rows.                    | 0       |\n",
+    "| skipna       | bool | not yet implemented                                           | True    |\n",
+    "| numeric_only | bool | Only use columns of the table that are of a numeric data type.                   | False   |\n",
+    "| ddof    | int  | Delta Degrees of Freedom. The divisor used in calculations is N - ddof, where N represents the number of elements. | 1 |\n",
+    "\n",
+    "**Returns:**\n",
+    "\n",
+    "| Type               | Description                                                          |\n",
+    "| :----------------: | :------------------------------------------------------------------- |\n",
+    "| Table         | The std across each row / column with the key corresponding to the row number or column name. |"
```
Comment (nipsn) [inline docs/user-guide/advanced/Pandas_API.ipynb:674]: Change both Type and Description. We are returning a Dictionary (`pykx.wrappers.Dictionary`) and we are not calculating the std.

**Review by nipsn — CHANGES_REQUESTED** [review]
Overall looking good, but needs some changes on the documentation side.

File: docs/user-guide/advanced/Pandas_API.ipynb (hunk 2 above)
Comment (tortolavivo23) [reply docs/user-guide/advanced/Pandas_API.ipynb]: 👍

File: docs/user-guide/advanced/Pandas_API.ipynb (hunk 1 above)
Comment (tortolavivo23) [reply docs/user-guide/advanced/Pandas_API.ipynb]: 👍

File: src/pykx/pandas_api/pandas_meta.py [hunk 3]
```diff
@@ -203,6 +203,30 @@ def mode(self, axis: int = 0, numeric_only: bool = False, dropna: bool = True):
...
+        if 'Keyed' in str(type(tab)):
+            tab = q.value(tab)
+        if numeric_only:
+            tab = _get_numeric_only_subtable(tab)
+
+        key_str = '' if axis == 0 else '`$string '
+        val_str = '' if axis == 0 else '"f"$value '
+        query_str = 'cols[tab]' if axis == 0 else 'til[count[tab]]'
+        where_str = ' where not (::)~/:r[;1]'
+        x_sed = f'{{(dev x) % sqrt count[x]-{ddof}}}'
```
[the remark this answers is not in the document]
Comment (tortolavivo23) [reply src/pykx/pandas_api/pandas_meta.py]: 👍

File: tests/test_pandas_api.py [hunk 4]
```diff
@@ -1489,6 +1489,89 @@ def test_df_sample(kx, q):
...
+            'a': [1, 2, 2, 4],
+            'b': [1, 2, 6, 7],
+            'c': [7, 8, 9, 10],
+            'd': [7, 11, 14, 14]
+        }
+    )
+    tab = kx.toq(df)
+    p_m = df.sem()
+    q_m = tab.sem()
+    assert all([p_m[c] == pytest.approx(q_m[c].py(),  1e-16)
```
Comment (nipsn) [inline tests/test_pandas_api.py:1504]: Minor: The value `1e-16` keeps being used throughout the tests. Maybe we could keep it in a variable `precision` or something similar. What do you think.

File: tests/test_pandas_api.py [hunk 5]
```diff
@@ -1489,6 +1489,89 @@ def test_df_sample(kx, q):
...
+    with pytest.raises(kx.QError):
+        q_m = tab.sem()
+    with pytest.raises(kx.QError):
+        q_m = tab.sem(axis=1)
+
+    df = pd.DataFrame(
+        {
+            'a': [1],
+        }
+    )
```
Comment (nipsn) [inline tests/test_pandas_api.py:1562]: This DataFrame declaration looks a bit too much to me. I suggest declaring it like so: 

```pd.DataFrame({'a': [1]})```

Mainly to reduce noise. What do you think?

**Review by nipsn — APPROVED** [review]

**Review by MiguelGomezC — APPROVED** [review]

**Review by cperezln — APPROVED** [review]

**Review by marcosvm13 — APPROVED** [review]

**Pull request merged**